server/schedule/checker/rule_checker.go

```diff
@@ -0,0 +1,266 @@
+	if s == nil {
+		return false
+	}
+	stateFilter := filter.StoreStateFilter{ActionScope: "rule-checker", TransferLeader: true}
+	if stateFilter.Target(c.cluster, s) {
+		return false
+	}
+	for _, rf := range fit.RuleFits {
+		if (rf.Rule.Role == placement.Leader || rf.Rule.Role == placement.Voter) &&
+			placement.MatchLabelConstraints(s, rf.Rule.LabelConstraints) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *RuleChecker) fixBetterLocation(region *core.RegionInfo, fit *placement.RegionFit, rf *placement.RuleFit) (*operator.Operator, error) {
+	if len(rf.Rule.LocationLabels) == 0 || rf.Rule.Count <= 1 {
+		return nil, nil
+	}
+	stores := getRuleFitStores(c.cluster, rf)
+	s := selector.NewReplicaSelector(stores, rf.Rule.LocationLabels, filter.StoreStateFilter{ActionScope: "rule-checker", MoveRegion: true})
+	oldPeerStore := s.SelectSource(c.cluster, stores)
+	if oldPeerStore == nil {
+		return nil, nil
+	}
+	oldPeer := region.GetStorePeer(oldPeerStore.GetID())
+	newPeerStore := SelectStoreToReplacePeerByRule("rule-checker", c.cluster, region, fit, rf, oldPeer)
+	stores = getRuleFitStores(c.cluster, removePeerFromRuleFit(rf, oldPeer))
+	oldScore := core.DistinctScore(rf.Rule.LocationLabels, stores, oldPeerStore)
+	newScore := core.DistinctScore(rf.Rule.LocationLabels, stores, newPeerStore)
+	if newScore <= oldScore {
+		log.Debug("no better peer", zap.Uint64("region-id", region.GetID()), zap.Float64("new-score", newScore), zap.Float64("old-score", oldScore))
+		return nil, nil
+	}
+	newPeer := &metapb.Peer{StoreId: newPeerStore.GetID(), IsLearner: oldPeer.IsLearner}
+	return operator.CreateMovePeerOperator("move-to-better-location", c.cluster, region, operator.OpReplica, oldPeer.GetStoreId(), newPeer)
+}
+
+func (c *RuleChecker) fixOrphanPeers(region *core.RegionInfo, fit *placement.RegionFit) (*operator.Operator, error) {
+	if len(fit.OrphanPeers) == 0 {
+		return nil, nil
+	}
+	peer := fit.OrphanPeers[0]
+	return operator.CreateRemovePeerOperator("remove-orphan-peers", c.cluster, 0, region, peer.StoreId)
+}
+
+func (c *RuleChecker) isDownPeer(region *core.RegionInfo, peer *metapb.Peer) bool {
+	for _, stats := range region.GetDownPeers() {
+		if stats.GetPeer().GetId() != peer.GetId() {
+			continue
+		}
+		storeID := peer.GetStoreId()
+		store := c.cluster.GetStore(storeID)
+		if store == nil {
+			log.Warn("lost the store, maybe you are recovering the PD cluster", zap.Uint64("store-id", storeID))
+			return false
+		}
+		if store.DownTime() < c.cluster.GetMaxStoreDownTime() {
+			continue
+		}
+		if stats.GetDownSeconds() < uint64(c.cluster.GetMaxStoreDownTime().Seconds()) {
+			continue
+		}
+		return true
+	}
+	return false
+}
+
+func (c *RuleChecker) isOfflinePeer(region *core.RegionInfo, peer *metapb.Peer) bool {
+	store := c.cluster.GetStore(peer.GetStoreId())
+	if store == nil {
+		log.Warn("lost the store, maybe you are recovering the PD cluster", zap.Uint64("store-id", peer.StoreId))
+		return false
+	}
+	return !store.IsUp()
```

Comment:
prefer to use `IsOffline`

Reply:
Here we treat `tomestone` as offline too. (it happens when force bury a store before moing out all regions). `replicaChecker`'s behaviour is like this too: https://github.com/pingcap/pd/blob/4d65bbefbc6db6e92fee33caa97415274512757a/server/schedule/checker/replica_checker.go#L216-L218

Reply:
Yes, but this function named `isOfflinePeer` which may be a little bit confused.

Reply:
em. do you have better suggestion? (replicaChecker uses `checkOfflinePeer` as function name too...)

Reply:
I think IsOfflineOrTomestone looks too verbose 😢